packages/jaspr_lints/lib/src/assists/tree_assist.dart

```diff
@@ -0,0 +1,221 @@
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/source/source_range.dart';
+import 'package:analyzer_plugin/protocol/protocol_common.dart';
+import 'package:custom_lint_builder/custom_lint_builder.dart';
+
+import '../utils.dart';
+
+class TreeAssistProvider extends DartAssist {
+  @override
+  void run(
+    CustomLintResolver resolver,
+    ChangeReporter reporter,
+    CustomLintContext context,
+    SourceRange target,
+  ) {
+    context.registry.addInvocationExpression((node) {
+      if (!target.coveredBy(node.function.sourceRange)) {
+        return;
+      }
+      if (!isComponentType(node.staticType)) {
+        return;
+      }
+      final indent = getLineIndent(resolver.lineInfo, node);
+      wrapComponent(resolver, reporter, node, indent);
+      removeComponent(resolver, reporter, node, indent, node.argumentList);
+      extractComponent(resolver, reporter, node, indent);
+    });
+    context.registry.addInstanceCreationExpression((node) {
+      if (!target.coveredBy(node.constructorName.sourceRange)) {
+        return;
+      }
+      if (!isComponentType(node.staticType)) {
+        return;
+      }
+      final indent = getLineIndent(resolver.lineInfo, node);
+      wrapComponent(resolver, reporter, node, indent);
+      removeComponent(resolver, reporter, node, indent, node.argumentList);
+      extractComponent(resolver, reporter, node, indent);
+    });
+  }
+
+  void wrapComponent(CustomLintResolver resolver, ChangeReporter reporter, Expression node, int lineIndent) {
+    var content = resolver.source.contents.data;
+    var lines = content.substring(node.offset, node.end).split('\n');
+
+    var htmlSource =
+        '([\n${''.padLeft(lineIndent)}${lines.map((s) => '  $s').join('\n')},\n${''.padLeft(lineIndent)}])';
+
+    void wrapWith(String name, [List<String>? suggestions]) {
+      final cb = reporter.createChangeBuilder(
+        priority: 2,
+        message: 'Wrap with $name',
+      );
+
+      cb.addDartFileEdit((builder) {
+        builder.addReplacement(node.sourceRange, (edit) {
+          if (suggestions != null) {
+            edit.addSimpleLinkedEdit(name, suggestions.first,
+                kind: LinkedEditSuggestionKind.METHOD, suggestions: suggestions);
+          } else {
+            edit.write(name);
+          }
+          edit.write(htmlSource);
+        });
+      });
+    }
+
+    wrapWith('html...', ['div', 'section', 'ul', 'li', 'h1', 'p', 'span', 'button']);
+    wrapWith('div');
+    wrapWith('section');
+    wrapWith('ul');
+    wrapWith('li');
+
+    final cb = reporter.createChangeBuilder(
```

Comment:
Do you thing that `cb` variable name is ok?
What about: `componentWrapperBuilder`?

Reply:
Done